packages/realm-react/src/__tests__/useProgress.test.tsx

```diff
@@ -0,0 +1,156 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import React, { PropsWithChildren } from "react";
+import {
+  EstimateProgressNotificationCallback,
+  ProgressDirection,
+  ProgressMode,
+  ProgressNotificationCallback,
+} from "realm";
+import { render } from "@testing-library/react-native";
+
+import { callMockedProgressNotifications, mockSyncedRealm } from "./mocks";
+import { RealmProvider, useProgress } from "..";
+import { sleep } from "./helpers";
+import { Text } from "react-native";
+
+const expectedProgress = {
+  [ProgressMode.ForCurrentlyOutstandingWork]: {
+    [ProgressDirection.Download]: [1, 0.2, 0.6, 1],
+    [ProgressDirection.Upload]: [1, 0.3, 0.7, 1],
+  },
+  [ProgressMode.ReportIndefinitely]: {
+    [ProgressDirection.Download]: [0, 0.25, 0.65, 1],
+    [ProgressDirection.Upload]: [0, 0.35, 0.75, 1],
+  },
+};
+
+const progressTestDuration = 100;
+
+const mockSyncedRealmWithProgress = () => {
+  const progressNotifiers: Map<ProgressNotificationCallback, NodeJS.Timeout> = new Map();
+  return mockSyncedRealm({
+    syncSession: {
+      addProgressNotification(direction, mode, callback) {
+        progressNotifiers.set(
+          callback,
+          callMockedProgressNotifications(
+            callback as EstimateProgressNotificationCallback,
+            // Make download progress "quicker" to compare different testing cases.
+            direction == ProgressDirection.Download ? progressTestDuration - 10 : progressTestDuration,
+            expectedProgress[mode][direction],
+          ),
+        );
+      },
+      removeProgressNotification: (callback) => {
+        clearInterval(progressNotifiers.get(callback));
+        progressNotifiers.delete(callback);
+      },
+    },
+  });
+};
+
+describe("useProgress", () => {
+  describe("all methods are callable and report a state", () => {
+    afterEach(() => {
+      jest.restoreAllMocks();
+    });
+
+    (
+      [
+        [ProgressMode.ReportIndefinitely, ProgressDirection.Download],
+        [ProgressMode.ReportIndefinitely, ProgressDirection.Upload],
+        [ProgressMode.ForCurrentlyOutstandingWork, ProgressDirection.Download],
+        [ProgressMode.ForCurrentlyOutstandingWork, ProgressDirection.Upload],
+      ] as [ProgressMode, ProgressDirection][]
```

Comment:
I'm sure this could be written without the type assertion 🤞

Reply:
unfortunately not, it thinks of itself as `(ProgressMode | ProgressDirection)[]`, I wish typescript was a little smarter.

Reply:
An `as const` seems simple enough: https://www.typescriptlang.org/play/?#code/KYOwrgtgBACgTgewOZ2AZzQWQQE2FAbwCgooAlYABwTgBcBJEPAMwEsRXbgAbATwBoSUAGI0AwmDioQtPgHkwtNLQCGTdkgDqNANaCAvkVCRYiFOjQARVqgDGtVghCEhlhAHcQ3BCpyDSAKqU3r4GRES2TspQAG4q3GDoUAC8UADaRAD0maSkAHoA-EJp8MioGNh4AHQU1HSMLOycPAKmZRbWdg5OVW6eITgAuv7ppeYVuMA1VDQMTMBsHFx8-G3jVjbA9o4gVUEDw8Vj5ViTVaJwElKgsrwKSqrqIFq6q8cdm9s9fV4+QyMlMwnSpTC5XaS3e7KNQ4DTaOB6NYnTpbbq7fZ-Q6DKAqNBQSIgZQAbiIQA

Reply:
Oh cool! I remember using `as const` before but dind't think of this in that scenario